pkg/raft/rafttype/util.go

```diff
@@ -0,0 +1,211 @@
+
+// Priority specifies per-entry priorities, that are local to the interaction
+// between a leader-replica pair, i.e., they are not an invariant of a
+// particular entry in the raft log (the replica could be the leader itself or
+// a follower). There are four priorities, ranging from low to high. These
+// form the foundation for a priority based admission control sub-system, for
+// the replication layer, where the functionality is split between the Raft
+// layer and higher layer that exercises Raft. We discuss Raft's
+// responsibility in the functionality below.
+//
+// Raft is not concerned with how the higher layer at the leader assigns
+// priorities, or communicates them to a replica.
+//
+// Raft provides (a) tracking of an Admitted vector (one element per priority,
+// for each replica), (b) pinging for followers (using MsgApps) when Admitted
+// vector elements are lagging behind Match, (c) (for followers) piggy-backing
+// Admitted vectors on MsgApp and MsgAppResp to allow the leader to converge
+// to the follower's state, (d) method for a replica to advance the value of
+// the Admitted vector.
+//
+// (d) is the entry point for the higher layer to participate in the liveness
+// of Admitted, but the nitty-gritty details of liveness are handled by Raft.
+//
+// Note that even though the priorities are per-entry on a leader-replica
+// pair, we expect the higher layer to advance Admitted for all priorities.
+// That is, if Admitted[LowPri]=10 and entries 11, 12 are assigned HighPri, it
+// is the responsibility of the higher layer to set Admitted[LowPri]=12,
+// without waiting for future LowPri entries to arrive.
+type Priority uint8
+
+const (
+	LowPri Priority = iota
+	NormalPri
+	AboveNormalPri
+	HighPri
+	NumPriorities
+)
+
+func (p Priority) String() string {
+	return redact.StringWithoutMarkers(p)
+}
+
+// SafeFormat implements the redact.SafeFormatter interface.
+func (p Priority) SafeFormat(w redact.SafePrinter, _ rune) {
+	switch p {
+	case LowPri:
+		w.Printf("LowPri")
+	case NormalPri:
+		w.Printf("NormalPri")
+	case AboveNormalPri:
+		w.Printf("AboveNormalPri")
+	case HighPri:
+		w.Printf("HighPri")
+	default:
+		panic("invalid raft priority")
+	}
+}
+
+func ConvertMessage(message raftpb.Message) Message {
+	return Message{
+		Type:       MessageType(message.Type),
+		To:         PeerID(message.To),
+		From:       PeerID(message.From),
+		Term:       message.Term,
+		LogTerm:    message.LogTerm,
+		Index:      message.Index,
+		Entries:    convertEntries(message.Entries),
+		Commit:     message.Commit,
+		Snapshot:   convertSnapshot(message.Snapshot),
+		Reject:     message.Reject,
+		RejectHint: message.RejectHint,
+		Context:    message.Context,
+	}
+}
+
+func convertEntries(entries []raftpb.Entry) []Entry {
+	result := make([]Entry, len(entries))
+	for i, entry := range entries {
+		result[i] = Entry{
+			Term:  entry.Term,
+			Index: entry.Index,
+			Type:  EntryType(entry.Type),
+			Data:  entry.Data,
+		}
+	}
```

Comment:
This causes an extra allocation on every message containing entries (`MsgProp`, `MsgApp`, etc)? Also, the `convertSnapshot`.

We would probably want to make these messages iterable or zero-copy somehow, so that they could be stepped into raft without an extra allocation.